meeshkan/serve/mock/data/entity.py

```diff
@@ -0,0 +1,121 @@
+import typing
+import uuid
+
+from http_types import Request
+from jsonpath_rw import parse
+from meeshkan.build.paths import _match_to_path
+from openapi_typed_2 import PathItem, Operation
+
+
+class Entity:
+    methods = ['get', 'put', 'post', 'delete', 'options', 'head', 'patch', 'trace']
+
+    def __init__(self, path: str, path_item: PathItem):
+        self._path = path
+        self._path_item = path_item
+        self._name = path_item._x["x-meeshkan-entity"]
+        self._id_path = parse(path_item._x["x-meeshkan-id-path"])
+        self._data = dict()
+        self._request_entity_selectors = self._build_request_entity_selectors(path_item)
+
+    def query(self, request: Request):
+        return [x for x in self._data.values() if self._filter(request)]
+
+    def query_one(self, request: Request):
+        id = self._get_id(request)
+        return self._data.get(id, {}) if id is not None else {}
+
+    def _get_id(self, request: Request):
+        id = _match_to_path(self._path, request.pathname).get('id', self._id_filter(request))
+        return id
+
+    def insert_from_request(self, request: Request):
+        id = self._get_id(request)
+        if id is None:
+            id = self._generate_id()
+        entity_val = self._request_entity_selectors[request.method].find()
+        entity_val = self._id_path.set(entity_val, id)
+        self._data[id] = entity_val
+        return entity_val
+
+    def insert(self, name, entity):
+        id = self._id_path.find(entity).value()
+        self._data[id] = entity
+
+    def upsert_from_request(self, request: Request):
+        id = self._get_id(request)
+        if id is None:
+            return self.insert_from_request(request)
+        else:
+            entity_val = self._request_entity_selectors[request.method].find()
+            return self._merge(self._data[id], entity_val)
+
+
+    def _merge(self, entity1, entity2):
+        for k, v in entity2.items():
+            entity1[k] = v
+        return entity1
+
+    def _extact_id(self, entity):
+        return self._id_path.find(entity)
+
+    def _filter(self, request: Request):
+        return lambda x: True
+
+    def _id_filter(self, request: Request):
+        return None
+
+    def __getitem__(self, key):
+        return self._data[key]
+
+    def __contains__(self, key):
+        return key in self._data
+
+    def __delitem__(self, key):
+        del self._data[key]
+
+    def __setitem__(self, key, value):
+        self._data[key] = value
+
+    def get(self, key, default=None):
+        return self._data.get(key, default)
+
+    def _build_request_entity_selectors(self, path_item: PathItem) -> typing.Mapping[str, typing.Callable]:
```

Comment:
A good place to consider lenses if you're interested in checking them out! They avoid this type of pyramid. See a good explanation in https://github.com/gcanti/monocle-ts, which takes a pyramid and flattens it using a lens. Not mission critical though.